src/controller/CommissionerDiscoveryController.cpp

```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
     }
     client->SetUDCClientProcessingState(UDCClientProcessingState::kObtainingOnboardingPayload);
 
-    if (mPasscodeService != nullptr)
+    if (mPasscodeService == nullptr)
     {
-        char rotatingIdString[chip::Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
-        Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
-                                            sizeof(rotatingIdString));
-        // Encoding::BytesToUppercaseHexString(client->GetRotatingId(), chip::Dnssd::kMaxRotatingIdLen, rotatingIdString,
-        //                                     sizeof(rotatingIdString));
+        HandleContentAppPasscodeResponse(0);
+        return;
+    }
+
+    ChipLogError(AppServer, "UX Ok: moving out of main thread");
+
+    // TODO: move this an external Thread
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackExternalWorkerOk, static_cast<void *>(client));
+
+    ChipLogError(AppServer, "UX Ok: done moving out of main thread");
+}
+
+/// Callback for getting execution into the main chip thread
+void CallbackHandleTargetContentAppCheck(System::Layer * aSystemLayer, void * aAppState)
+{
+    TargetAppInfo * target = static_cast<TargetAppInfo *>(aAppState);
+    GetCommissionerDiscoveryController()->InternalHandleTargetContentAppCheck(*target, target->passcode);
+    delete target;
+}
+
+void CommissionerDiscoveryController::HandleTargetContentAppCheck(TargetAppInfo target, uint32_t passcode)
+{
+    // need to ensure callback is on main chip thread
+    target.passcode      = passcode;
+    TargetAppInfo * info = new TargetAppInfo(target);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackHandleTargetContentAppCheck, info);
+}
+
+void CommissionerDiscoveryController::InternalHandleTargetContentAppCheck(TargetAppInfo target, uint32_t passcode)
+{
+    CheckValidSession();
```

Comment:
Should `assertChipStackLockedByCurrentThread()`

Reply:
added this assert